src/trans.rs

```diff
@@ -226,9 +265,42 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
                     let expr = unsafe { BinaryenReturn(self.module, expr) };
                     binaryen_stmts.push(expr);
                 }
+                TerminatorKind::Switch { ref discr, .. } => {
+                    let adt = self.trans_lval(discr);
+                    let adt_ty = self.mir.lvalue_ty(*self.tcx, discr).to_ty(*self.tcx);
+
+                    if adt.offset.is_some() {
+                        panic!("unimplemented Switch with offset");
+                    }
+
+                    let adt_layout = self.type_layout(adt_ty);
+                    let discr_val = match *adt_layout {
+                        Layout::General { discr, .. } => {
+                            let discr_size = discr.size().bytes() as u32;
+                            debug!("emitting GetLocal({}) + Load for ADT Switch condition", adt.index.0);
+                            unsafe {
+                                let ptr = BinaryenGetLocal(self.module, adt.index, BinaryenInt32());
+                                BinaryenLoad(self.module, discr_size, 0, 0, 0, BinaryenInt32(), ptr)
+                            }
+                        }
+                        Layout::CEnum { .. } => {
+                            debug!("emitting GetLocal({}) for CEnum Switch condition", adt.index.0);
+                            unsafe {
+                                BinaryenGetLocal(self.module, adt.index, BinaryenInt32())
+                            }
+                        }
+                        _ => panic!("unimplemented discrimant value for Layout {:?}", adt_layout)
+                    };
+
+                    block_kind = BinaryenBlockKind::Switch(discr_val);
+                }
                 TerminatorKind::Call {
                     ref func, ref args, ref destination, ..
                 } => unsafe {
+                    // NOTE: plan for the calling convention: i32/i64 f32/f64 are to be passed using the wasm stack and function parameters.
+                    //       For the other types, the manual stack in linear memory will be used, and pointers into this stack
+                    //       passed as i32s. A call to a function returning a struct will require preparing the output return value space on
+                    //       the caller function's frame, and the called function will write its return value there to avoid memcpys
```

Comment:
but to answer your question in advance, I was thinking: we could pass the output pointer, but would you see an issue if we just write to the frame pointer ?